### DISCUSSION TITLE
Conversion to HLSL

### DISCUSSION THREAD

**Discussion by domportera** [stegu/psrdnoise#10] [Show and tell]
Hey all, I've converted this and the [webgl-noise](https://github.com/stegu/webgl-noise) repos to HLSL. Thank you stegu for this amazing work, I look forward to using it for years to come, and I hope my translation helps others do the same.

**Comment by chmodseven** [comment]
> Hey all, I've converted this and the [webgl-noise](https://github.com/stegu/webgl-noise) repos to HLSL. Thank you stegu for this amazing work, I look forward to using it for years to come, and I hope my translation helps others do the same.

Nice work. I've also done and repo'd a port of the psrdnoise GLSL and its various example permutations to Unity shaders and ShaderGraph nodes. It's certainly a wonderful gift that stegu has given us, even if I don't understand word one of the maths involved lol.

**Reply by domportera** [reply]
amazing!! thank you for this, I was not looking forward to having to work this out myself. starred :)

yeah all this math highly specific math is waaaay over my head too, this man has tapped into the arcane lmao

fun fact: the -min variants seem to just be "minified" versions of the shaders, and based on my investigation should be precisely identical to their non-min variants. for our sake with hlsl/unity, probably unnecessary

**Reply by chmodseven** [reply]
> amazing!! thank you for this, I was not looking forward to having to work this out myself. starred :)
> 
> yeah all this math highly specific math is waaaay over my head too, this man has tapped into the arcane lmao
> 
> fun fact: the -min variants seem to just be "minified" versions of the shaders, and based on my investigation should be precisely identical to their non-min variants. for our sake with hlsl/unity, probably unnecessary

Heh, yeah I think I'd already gone and ported the min one before I realised it was identical to the verbose one, so I decided to keep it in anyway.  IIRC all the variants like fBM, tendrils etc hook into the verbose one for their common code, but that could just as easily be swapped out to the min, although like you say it's probably unnecessary once the overall shader is compiled anyway.

As for combining all our stuff via a PR, that would be great, however my knowledge when it comes to git merging is shamefully inadequate, so I don't really know what to do there.  I'd be happy for one of you guys to copy-paste whatever is worth keeping out of mine to add into stegu's main repo, or whichever other method works best.  BTW I'd also be quite interested to see the workings of the compute shader usage you mentioned, if that's something you'd be able to add to your repo.

**Reply by domportera** [reply]
> Heh, yeah I think I'd already gone and ported the min one before I realised it was identical to the verbose one, so I decided to keep it in anyway
> 
heh I actually did the exact same thing, but just deleted it in a later commit 😅

> As for combining all our stuff via a PR, that would be great, however my knowledge when it comes to git merging is shamefully inadequate, so I don't really know what to do there. I'd be happy for one of you guys to copy-paste whatever is worth keeping out of mine to add into stegu's main repo, or whichever other method works best. BTW I'd also be quite interested to see the workings of the compute shader usage you mentioned, if that's something you'd be able to add to your repo.

I'm not sure how stegu would feel about something that's heavily dependent on a specific software platform (unity), so that's totally up to him. However if there's anything we can do to combine our powers in our own repos that sounds good to me :)

[this repo](https://github.com/domportera/unity-perlin-noise-on-gpu/tree/main/Runtime) is woefully underworked but I plan on actually developing it. just kinda pushed it up to make sure it survives a potential apocalyptic event. I did that little compute shader (thanks much to something i found online and adapted/optimized which i've since lost track of - eventually google searches of that source led me to the REAL source - the ashima repo) with just an isolated cnoise, but it does work with one of the (identical) simple perlin 4D noises.

it should probably be optimized to just use 2 or 3D but that's a problem for another day - i was just getting my feet wet. but it works more or less exactly how you'd expect it would in a ShaderLab shader, just much simpler! an include at the top (or theoretically anywhere before where you call it) and then you just call the function like you would expect.

the C# code I use to run my compute shaders is actually a pretty involved compute shader pipeline I've been developing at work, but unfortunately due to the fact that I've been doing it at work, it's closed-source against my will 😢

**Comment by stegu** [comment]
I would be happy to include your ports in my repo. There is already a WGSL
port there, because I felt like learning it. Just submit a pull request,
and I will accept it.
Because there are multiple ports, and because they are external
contributions, I would suggest submitting them as separate subfolders. I
might take a stab at merging the versions some time, but for now, it would
be a great addition to just have at least one working HLSL port in there,
even if it wouldn't be actively maintained by me and tested only by others.
People are asking me for it, so there is most definitively an interest out
there. My own reluctance to get up close and personal with HLSL (and
Direct3D) doesn't make me oblivious to the fact that it exists and is being
used.

**Reply by domportera** [reply]
id love to submit a PR, I just want to warn you most of this is untested 😅 though it seems like it should work, and having other more talented people test it and find (hypothetical, surely non-existent) problems would be useful for me too.

I understand the reluctance to get into Direct3D, but Vulkan supports HLSL now :)

how exactly would you like me today integrate this? only the psrd shaders, or the whole bunch? could also potentially take a git submodule approach but that might be more trouble than it's worth

**Comment by stegu** [comment]
Yeah, sorry for not being more clear, the "-min" versions are the same
functions, only with more compact source code. Because GLSL shaders are
compiled from source text by the application, people who write demos don't
care for comments
and verbose license texts.
If you're not itching to make a pull request, you can simply send me
whatever code you want to share in a zip file, or send me a link. My e-mail
address is "stefan dot gustavson at gmail dot com", as listed in the source
for psrdnoise. I probably won't be able to test HLSL stuff any time soon,
but hosting it on Github makes it easy for others to submit fixes if they
find a bug.

> this man has tapped into the arcane

FYI, I'm not actually a wizard, just a mad scientist. (Insert diabolical
"Muhahahaa", picture me in a double-buttoned lab coat and weird goggles, in
a room lit by sparks.)

**Reply by domportera** [reply]
I don't mind making a pull request! I'll PR everything and you can select what you like/give feedback on how you'd prefer it being structured in your repository

>FYI, I'm not actually a wizard, just a mad scientist. (Insert diabolical
>"Muhahahaa", picture me in a double-buttoned lab coat and weird goggles, in
>a room lit by sparks.)

LMAO idk man - 1000 years ago you could be burned at the stake for that so i'll hold onto my suspicions 😉

**Comment by stegu** [comment]
I'm looking forward to any PRs or links, or file dumps over email. I am
struggling with a pretty bad case of post-Covid at the moment, with an
often overwhelming brain fatigue, so I might not be quick to act, but
please don't take that as a lack of interest.

Myself, I recently modified a plain C software version of psrdnoise2 to use
only integer math and mostly 8-bit data types, to run on the lower ranks of
Arduino boards. It's fast enough for low-res LED effects (in my case, a
Neopixel light saber blade) and sound synthesis (also for that light
saber). That code might find a place in the repo as well if I include
"miscellaneous ports".

**Reply by domportera** [reply]
damn feel better! take all the time you need, i appreciate the heads up. I'll get to it at my soonest convenience. I'll also probably make another PR at a later date translating your wonderful webGL examples

WOW i can name a few people who would love something like that, nice work!!

**Comment by stegu** [comment]
Shannon, your hard and appreciated work with the Unity shader ports is now
at least linked and mentioned from stegu/psrdnoise. Given my current
less-than-ideal state of mind, I'm not sure I would get it right if I tried
to decide what to include if I were to duplicate your files in my repo, but
feel free to put together a pull request if you want your files to be
duplicated in, say, a directory named "src/unity".

**Reply by chmodseven** [reply]
> Shannon, your hard and appreciated work with the Unity shader ports is now at least linked and mentioned from stegu/psrdnoise. Given my current less-than-ideal state of mind, I'm not sure I would get it right if I tried to decide what to include if I were to duplicate your files in my repo, but feel free to put together a pull request if you want your files to be duplicated in, say, a directory named "src/unity".

Thanks Stefan, I appreciate it and hope you're feeling better soon - I get brain fog sometimes from some medication side effects, and it's no fun at all, I know!

I think the link alone should be sufficient for Unity devs to find what they want, so I don't think we need to bother with a pull request.

**Comment by Ollhax** [comment]
Hey there! I'm just tapping into the [HLSL ports](https://github.com/stegu/psrdnoise/blob/main/src/hlsl/stegu-psrdnoise/psrdnoise2.hlsl) and I think I found an error. This part:

```
float3 psi = hash * 0.07482 + alpha;
float cospsi = cos(psi); // Error converting float3 -> float1
float sinpsi = sin(psi); // Same
float3 gx = float3(cospsi, cospsi, cospsi);
float3 gy = float3(sinpsi, sinpsi, sinpsi);
```

This doesn't compile for me, but not entirely sure it's an error or if it's because I'm using [slang](https://shader-slang.org/) for my shaders which might not map 100% to HLSL. Here's the [original](https://github.com/stegu/psrdnoise/blob/main/src/psrdnoise2.glsl) for comparison:

```
vec3 psi = hash * 0.07482 + alpha;
vec3 gx = cos(psi);
vec3 gy = sin(psi);
```

I ended up using this, which also is probably a bit more optimal because it calculates sin and cos at the same time:

```
float3 psi = hash * 0.07482 + alpha;
float3 gx;
float3 gy;
sincos(psi, gy, gx);
```

**Comment by stegu** [comment]
I didn't do the HLSL ports, and I have no easy means for testing this, but
that part does look wrong. The hash is different for each grid point, and
the ported code doesn't look as if it handles that properly, with or
without your fix.

**Comment by stegu** [comment]
I'm sorry for my confusion, your fix seems fine. The original code is
nonsense, though, and doesn't look like it ever worked.

**Comment by domportera** [comment]
@Ollhax @stegu I didn't personally get to test every bit of this, as I really only needed a part of this noise library and just converted the rest while I was there as well. If something looks incorrect, it probably is, and your input here is invaluable! glad you caught this. I dont have time at the moment and am liable to forget, but this fix should definitely be upstreamed

it should be worth noting that the float conversion issues might be compiler-specific, as it did compile on my end using Unity's compiler, though that is far from a proper reference compiler that this conversion should hold itself accountable to
